### ISSUE TITLE
Belos build issue when Tpetra is disabled

### ISSUE THREAD

**Issue by jwillenbring** [trilinos/Trilinos#89]
@pbosler 

I was trying to build Belos as part of a new nightly no C++11 build, and it failed in doc/ due to Tpetra not being enabled. Tpetra is an optional dependence for Belos. Here is the error output:

[ 95%] Building CXX object packages/belos/doc/parameterList/CMakeFiles/Belos_ValidParameters.dir/createValidParameterList.cpp.o
/home/jmwille/TrilinosTest/Trilinos/packages/belos/doc/parameterList/createValidParameterList.cpp:66:34: error: Tpetra_MultiVector.hpp: No such file or directory
In file included from /home/jmwille/TrilinosTest/Trilinos/packages/belos/doc/parameterList/createValidParameterList.cpp:67:
/home/jmwille/TrilinosTest/Trilinos/packages/belos/doc/parameterList/../../tpetra/src/BelosTpetraAdapter.hpp:83:31: error: Tpetra_Operator.hpp: No such file or directory
In file included from /home/jmwille/TrilinosTest/Trilinos/packages/belos/doc/parameterList/createValidParameterList.cpp:67:
/home/jmwille/TrilinosTest/Trilinos/packages/belos/doc/parameterList/../../tpetra/src/BelosTpetraAdapter.hpp:114: error: ‘::Tpetra’ has not been declared
...

This appears to be due to changes made on Nov 17. Here is the critical subset of my cmake arguments:

-DTrilinos_ENABLE_CXX11=OFF \
-D Trilinos_ENABLE_TESTS:BOOL=ON \
-DCMAKE_C_COMPILER=/usr/lib64/openmpi/bin/mpicc \
-DCMAKE_CXX_COMPILER=/usr/lib64/openmpi/bin/mpic++ \
-DCMAKE_Fortran_COMPILER=/usr/lib64/openmpi/bin/mpif77 \
-DTrilinos_ENABLE_NOX=ON \
   -DNOX_ENABLE_LOCA=ON \
-DTrilinos_ENABLE_EpetraExt=ON \
   -DEpetraExt_BUILD_BTF=ON \
-DTrilinos_ENABLE_TrilinosCouplings=ON \
-DTrilinos_ENABLE_Ifpack=ON \
-DTrilinos_ENABLE_Isorropia=ON \
-DTrilinos_ENABLE_AztecOO=ON \
-DTrilinos_ENABLE_Belos=ON \
-DTrilinos_ENABLE_Teuchos=ON \
-DTeuchos_ENABLE_COMPLEX=ON \
-DTrilinos_ENABLE_Amesos=ON \
   -DAmesos_ENABLE_KLU=ON \
-DTrilinos_ENABLE_Sacado=ON \
-DTrilinos_ENABLE_Zoltan=ON \
-DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES=OFF \
-DTPL_ENABLE_MPI=ON \

I think it would be possible to basically remove all of the package enables except Belos. You could just ifdef the whole directory essentially if Tpetra isn't enabled, but I wanted to give you the opportunity to be more selective than that.

**Comment by mhoemmen** [comment]
This looks like a Bill Spotz and Dena Vigil thing... Could someone poke them?

mfh

On Jan 21, 2016, at 10:48 PM, jwillenbring <notifications@github.com<mailto:notifications@github.com>> wrote:


@pbosler<https://github.com/pbosler>

I was trying to build Belos as part of a new nightly no C++11 build, and it failed in doc/ due to Tpetra not being enabled Tpetra is an optional dependence for Belos Here is the error output:

[ 95%] Building CXX object packages/belos/doc/parameterList/CMakeFiles/Belos_ValidParametersdir/createValidParameterListcppo
/home/jmwille/TrilinosTest/Trilinos/packages/belos/doc/parameterList/createValidParameterListcpp:66:34: error: Tpetra_MultiVectorhpp: No such file or directory
In file included from /home/jmwille/TrilinosTest/Trilinos/packages/belos/doc/parameterList/createValidParameterListcpp:67:
/home/jmwille/TrilinosTest/Trilinos/packages/belos/doc/parameterList///tpetra/src/BelosTpetraAdapterhpp:83:31: error: Tpetra_Operatorhpp: No such file or directory
In file included from /home/jmwille/TrilinosTest/Trilinos/packages/belos/doc/parameterList/createValidParameterListcpp:67:
/home/jmwille/TrilinosTest/Trilinos/packages/belos/doc/parameterList///tpetra/src/BelosTpetraAdapterhpp:114: error: '::Tpetra' has not been declared

This appears to be due to changes made on Nov 17 Here is the critical subset of my cmake arguments:

-DTrilinos_ENABLE_CXX11=OFF \
-D Trilinos_ENABLE_TESTS:BOOL=ON \
-DCMAKE_C_COMPILER=/usr/lib64/openmpi/bin/mpicc \
-DCMAKE_CXX_COMPILER=/usr/lib64/openmpi/bin/mpic++ \
-DCMAKE_Fortran_COMPILER=/usr/lib64/openmpi/bin/mpif77 \
-DTrilinos_ENABLE_NOX=ON \
-DNOX_ENABLE_LOCA=ON \
-DTrilinos_ENABLE_EpetraExt=ON \
-DEpetraExt_BUILD_BTF=ON \
-DTrilinos_ENABLE_TrilinosCouplings=ON \
-DTrilinos_ENABLE_Ifpack=ON \
-DTrilinos_ENABLE_Isorropia=ON \
-DTrilinos_ENABLE_AztecOO=ON \
-DTrilinos_ENABLE_Belos=ON \
-DTrilinos_ENABLE_Teuchos=ON \
-DTeuchos_ENABLE_COMPLEX=ON \
-DTrilinos_ENABLE_Amesos=ON \
-DAmesos_ENABLE_KLU=ON \
-DTrilinos_ENABLE_Sacado=ON \
-DTrilinos_ENABLE_Zoltan=ON \
-DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES=OFF \
-DTPL_ENABLE_MPI=ON \

I think it would be possible to basically remove all of the package enables except Belos You could just ifdef the whole directory essentially if Tpetra isn't enabled, but I wanted to give you the opportunity to be more selective than that

-

**Comment by jwillenbring** [comment]
@wfspotz 
Mark suggested including Bill and Dena on this ticket. Since Dena doesn't seem to have an account set up for accessing Trilinos yet, I just mentioned Bill.

**Comment by jwillenbring** [comment]
@hkthorn 
Mentioning Heidi since she is interested in the resolution of this.

**Comment by mhoemmen** [comment]
@wfspotz -- It looks like your ParameterList documentation generation thing depends on Tpetra.  That's OK, but if we disable Tpetra, would you be OK with your stuff getting disabled (in Belos)?  Thanks!

**Comment by wfspotz** [comment]
Logically, if Tpetra is disabled, and Belos is enabled, we should still be able to generate HTML documentation for Belos ...

**Comment by pbosler** [comment]
The executable in Belos that generates the .xml files for html display uses Tpetra.  If the xml files don't exist, the parameter lists won't display in the html but doxygen should still generate it I think.

**Comment by hkthorn** [comment]
Unfortunately, the code was written using Tpetra objects.  Since acquiring the documentation from the solver requires the creation of a solver, there has to be a declaration of the template objects (operator, vector, scalar).  The declaration was done using Tpetra.  A conditional build is performed, based on the whether Tpetra is being built, on the development branch and the same fix has been released in 12.6.3.

--

Heidi K. Thornquist
Electrical Models & Simulation
Sandia National Laboratories
Albuquerque, NM  87185-1323

**Comment by mhoemmen** [comment]
@hkthorn Does this mean you fixed it?  (Sorry, the passive voice confuses me a bit.)

**Comment by hkthorn** [comment]
Yes, I fixed the build error.  In doing so I disabled the documentation building when Tpetra is disabled.  If the documentation needs to be built regardless of linear algebra, the file should use the implementation of the linear algebra interface in Belos/test/MVOPTester that is always built.

So, yes I fixed the build error, but the fix may not be what is desired.

Thanks,
Heidi

--

Heidi K. Thornquist
Electrical Models & Simulation
Sandia National Laboratories
Albuquerque, NM  87185-1323
